server/src/main/java/io/spine/server/delivery/CatchUpStation.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * <b>3. Catch-up {@code COMPLETED}.</b>
+ *
+ * <p>At this stage the event history is fully processed. The inboxes contain the messages in
+ * {@code TO_CATCH_UP} status, which are in fact the mix of the last portion of the replayed event
+ * history and potentially some "paused" live events. So, the station dispatched all the matching
+ * messages in {@code TO_CATCH_UP} status.
+ *
+ * <p>If the deduplication window is {@linkplain DeliveryBuilder#setDeduplicationWindow(Duration)
+ * set in the system}, all the delivered messages are set to be kept in their storages
+ * for the duration, corresponding to the width of the window. In this way, they will become usable
+ * for the potential deduplication.
+ *
+ * <h1>Deduplication and reordering</h1>
+ *
+ * <p>Prior to the dispatching, the messages are deduplicated in scope of this message batch.
+ * Please note, that the deduplication window is NOT taken into the account, as the historical
+ * events may all have been delivered to their entities somewhen in the past.
+ *
+ * <p>All the duplicates are marked as such in the conveyor and are removed from their storage
+ * later.
+ *
+ * <p>Another change made before the actual dispatching is reordering of the messages. The messages
+ * are sorted chronologically, putting the events of a framework-internal {@link CatchUpStarted}
+ * type first. It allows to guarantee that the targets will know of the started catch-up before
+ * any message in {@code TO_CATCH_UP} status is dispatched to them.
+ *
+ * <p>The ordering changes are not reflected on the message order in the conveyor.
+ */
+final class CatchUpStation extends Station {
+
+    private static final Duration HOW_LONG_TO_KEEP = Durations.fromMillis(1000);
+    private static final Comparator<InboxMessage> COMPARATOR = new CatchUpMessageComparator();
+
+    private final DeliveryAction action;
+    private final Iterable<CatchUp> jobs;
+
+    CatchUpStation(DeliveryAction action, Iterable<CatchUp> jobs) {
+        super();
+        this.action = action;
+        this.jobs = jobs;
+    }
+
+    /**
+     * Processes the messages on the conveyor, delivering those sent for catch-up.
+     *
+     * @param conveyor
+     *         the conveyor on which the messages are travelling
+     * @return the result of the processing telling how many messages were dispatched and whether
+     *         there were any errors in that
+     */
+    @SuppressWarnings({"MethodWithMultipleLoops", "OverlyComplexMethod", "OverlyNestedMethod"})
+    @Override
+    public final Result process(Conveyor conveyor) {
+        Map<DispatchingId, InboxMessage> dispatchToCatchUp = new HashMap<>();
+
+        for (InboxMessage message : conveyor) {
+            for (CatchUp job : jobs) {
+                CatchUpStatus jobStatus = job.getStatus();
+
+                if (job.matches(message)) {
+                    DispatchingId dispatchingId = new DispatchingId(message);
+                    if (jobStatus == IN_PROGRESS) {
```

Comment:
@alexander-yevsyukov This piece is explained by the class-level doc. The same applies the `LiveDeliveryStation`.

This whole class _is_ itself a method object, as is the `LiveDeliveryStation`. 

I will try to extract pieces one more time. It's may make it look better designed. But the extra call chain will make it harder to follow.
And, to avoid creating extra objects, I will probably have to modify the variables by references. 

Let's see if it makes the code easier for non-authors :)